apps/discord-bot/src/triggers/cta-post.ts

```diff
@@ -0,0 +1,280 @@
+        const thread = await this.createCTAThread(msg)
+        if (thread) {
+          this.startCTAReactionCollector(msg, thread)
+        }
+      }
+    }
+  }
+
+  private async getFinishedReactions(msg: Message): Promise<object> {
+    const finishedReactions = msg.reactions.cache.filter(reaction => finishedEmoji === reaction.emoji.name);
+    const roleReactions: { [region: string]: string[] } = {};
+
+    finishedReactions.forEach(async (reaction) => {
+      const users = await reaction.users.fetch();
+      users.forEach(async (user) => {
+        if (user.id === msg.client.user.id) {
+          return;
+        }
+        const member = msg.guild?.members.fetch(user.id);
+        const memberRoles = (await member)?.roles.cache;
+        const role = memberRoles?.filter(r => regionRoles.includes(r.name));
+        const memberRegionRole = role?.first()?.name;
+
+        if (memberRegionRole != undefined) {
+          if (!roleReactions[memberRegionRole]) {
+            roleReactions[memberRegionRole] = [];
+          }
+          if (!roleReactions[memberRegionRole]?.includes(user.id)) {
+            roleReactions[memberRegionRole]?.push(user.id);
+            console.log(`Adding ${role?.first()?.name} member, ${user.displayName}`);
+          }
+        }
+      });
+    });
+    return roleReactions;
+  }
+
+  private async activeCTAThread(msg: Message): Promise<PublicThreadChannel | undefined> {
+    if (msg.hasThread && msg.thread?.type === ChannelType.PublicThread && !msg.thread?.archived) {
+      return msg.thread;
+    }
+  }
+
+  private async createCTAThread(msg: Message): Promise<PublicThreadChannel | undefined> {
+
+    if (msg.channel.type === ChannelType.GuildText) {
+      const thread = msg.startThread({
+        // set as the title of the message
+
+        name: 'CTA Completion by Region',
+        autoArchiveDuration: ThreadAutoArchiveDuration.OneWeek,
+        reason: 'Tracking CTA participation.',
+      });
+
+      const threadChannel = await thread;
+      threadChannel.send({
+        content: `Mark your completion of the CTA by reacting with ${finishedEmoji}.\n\n A chart will be posted in an hour, and updated every hour until the thread is archived, showing the number of completions per region.`,
+        flags: [MessageFlags.SuppressNotifications]
+      });
+      msg.react(finishedEmoji)
+
+      return thread
+    }
+
+    return undefined
+  }
+
+  // start a reaction collector on the given message
+  private async startCTAReactionCollector(msg: Message, thread: PublicThreadChannel): Promise<void> {
+    const filter = (reaction, user) => reaction.emoji.name === finishedEmoji && user.id != msg.client.user.id
+    const collector = msg.createReactionCollector({ filter, time: oneHour });
+    const roleReactions = await this.getFinishedReactions(msg)
+
+    collector.on('collect', async (reaction, user) => {
+      const member = msg.guild?.members.fetch(user.id)
+      if (member != undefined) {
```

Comment:
I also personally find code easier to digest when there are fewer levels of nesting. Instead of `if (variable != undefined)`, I prefer `if (variable === undefined) return`. This is functionally the same, but stops you from having to create another block.

In cases like this though, `member` really shouldn't be `undefined` unless someone made a reaction and somehow left the server super quickly afterwards. So an error could also be appropriate here (as long as throwing one won't prevent future runs of the collector.)